test/server/proxy-test.js

```diff
@@ -2016,31 +2016,43 @@ describe('Proxy', () => {
 
                 session.addRequestEventListeners(rule, {
                     onRequest: e => {
-                        expect(e.isAjax).to.be.false;
+                        return new Promise(resolve => {
+                            expect(e.isAjax).to.be.false;
 
-                        expect(e._requestInfo.url).eql(url);
-                        expect(e._requestInfo.method).eql('get');
-                        expect(e._requestInfo.requestId).to.be.not.empty;
-                        expect(e._requestInfo.sessionId).to.be.not.empty;
-                        expect(e._requestInfo.body.length).eql(0);
-                        expect(e._requestInfo.headers).to.include({ 'content-type': 'application/javascript; charset=utf-8' });
+                            expect(e._requestInfo.url).eql(url);
+                            expect(e._requestInfo.method).eql('get');
+                            expect(e._requestInfo.requestId).to.be.not.empty;
+                            expect(e._requestInfo.sessionId).to.be.not.empty;
+                            expect(e._requestInfo.body.length).eql(0);
+                            expect(e._requestInfo.headers).to.include({ 'content-type': 'application/javascript; charset=utf-8' });
 
-                        requestEventIsRaised = true;
+                            requestEventIsRaised = true;
+
+                            resolve();
+                        });
                     },
 
                     onConfigureResponse: e => {
-                        configureResponseEventIsRaised = true;
+                        return new Promise(resolve => {
+                            configureResponseEventIsRaised = true;
+
+                            e.opts.includeHeaders = true;
+                            e.opts.includeBody    = true;
 
-                        e.opts.includeHeaders = true;
-                        e.opts.includeBody    = true;
+                            resolve();
+                        });
                     },
 
                     onResponse: e => {
-                        expect(e.statusCode).eql(200);
-                        expect(e.headers).to.include({ 'content-type': 'application/javascript; charset=utf-8' });
-                        expect(e.body.toString()).eql(resourceContent);
+                        return new Promise(resolve => {
```

Comment:
I would like to add a delay (e.g. 100ms) before resolving test hook promises because promises can be resolved before event handlers and timers and there will be no differences between a resolved promise and synchronous code. https://jsblog.insiderattack.net/promises-next-ticks-and-immediates-nodejs-event-loop-part-3-9226cbe7a6aa